src/psygnal/_signal.py

```diff
@@ -944,22 +945,35 @@ def __call__(
             check_types=check_types,
         )
 
+    def _run_emit_loop_inner(self) -> None:
+        try:
+            with Signal._emitting(self):
+                q_len = len(self._emit_queue)
+                for i in range(q_len):
+                    args = self._emit_queue[i]
+                    for caller in self._slots:
+                        caller.cb(args)
+                self._emit_queue = self._emit_queue[q_len:]
+                if len(self._emit_queue) > 0:
+                    self._run_emit_loop_inner()
+        except RecursionError as e:
+            raise RecursionError(
+                f"RecursionError in {caller.slot_repr()} when "
+                f"emitting signal {self.name!r} with args {args}"
+            ) from e
+        except Exception as e:
+            raise EmitLoopError(cb=caller, args=args, exc=e, signal=self) from e
+
     def _run_emit_loop(self, args: tuple[Any, ...]) -> None:
         # allow receiver to query sender with Signal.current_emitter()
         with self._lock:
-            with Signal._emitting(self):
-                for caller in self._slots:
-                    try:
-                        caller.cb(args)
-                    except RecursionError as e:
-                        raise RecursionError(
-                            f"RecursionError in {caller.slot_repr()} when "
-                            f"emitting signal {self.name!r} with args {args}"
-                        ) from e
-                    except Exception as e:
-                        raise EmitLoopError(
-                            cb=caller, args=args, exc=e, signal=self
-                        ) from e
+            self._emit_queue.append(args)
+            if len(self._emit_queue) > 1:
+                return None
+            try:
+                self._run_emit_loop_inner()
+            finally:
+                self._emit_queue.clear()
 
         return None
```

Comment:
here is a single _run_emit_loop that I feel reads much more clearly than the current two-function method that does funny stuff with `self._emit_queue = self._emit_queue[q_len:]`.  It also passes all the tests and works fine with mypyc.

```python
    def _run_emit_loop(self, args: tuple[Any, ...]) -> None:
        # allow receiver to query sender with Signal.current_emitter()
        with self._lock:
            self._emit_queue.append(args)
            if Signal.current_emitter() is self:
                return
            try:
                with Signal._emitting(self):
                    count = 0
                    while self._emit_queue:
                        args = self._emit_queue.pop(0)
                        for caller in self._slots:
                            caller.cb(args)
                        count += 1
                        if count > sys.getrecursionlimit():
                            raise RecursionError
            except RecursionError as e:
                raise RecursionError(f"RecursionError in signal {self.name!r}") from e
            except Exception as e:
                raise EmitLoopError(cb=caller, args=args, exc=e, signal=self) from e
```

 `self._emit_queue` can easily be made a deque.  If you don't like this, or if you have concerns about thread safety.  Can you perhaps provide a test that fails with the code above, but not with your code?  Because otherwise, i find it much easier to reason about.